keystone_api/apps/allocations/tasks.py

```diff
@@ -0,0 +1,184 @@
+        # TODO: This may create historical usage that keystone doesnt know about if raw usage is not reset
+        #  for the account upon it actually being added into the application
+        set_cluster_limit(account_name, cluster.name, get_cluster_usage(account_name, cluster.name))
+        return
+
+    # TODO: plan to perform a reset of rawusage before deployment. This makes it so an initial usage does not need
+    #  to be tracked. Resetting rawusage will be part of the install procedure for us and others
+
+    # Gather allocations that have expired but do not have a final usage value
+    # (still contributing to current limit as active SUs instead of historical usage)
+    closing_allocations_query = (Allocation.objects.filter(request__group=account,
+                                                           cluster=cluster,
+                                                           request__approved=True,
+                                                           final=None,  # TODO: will this work?
+                                                           request__expire__lte=date.today())
+                                                   .order_by("request__expire"))
+
+    # Gather all allocations belonging to "active" (started today or before and approved) Allocation Requests
+    # for the account
+    active_allocations_query = (Allocation.objects.filter(request__group=account,
+                                                          cluster=cluster,
+                                                          request__approved=True,
+                                                          request__active__lte=date.today(),
+                                                          request__expire__gt=date.today())
+                                                  .order_by("request__expire"))
+
+    # Determine the SU contribution by active allocations
+    # TODO: Is this zero if there are no active allocations?
+    active_sus = active_allocations_query.aggregate(Sum("awarded"))
+
+    # Determine usage that can be covered:
+    # total usage on the cluster (from slurm) - historical usage (current limit from slurm - active SUs - closing SUs)
+    current_usage = (get_cluster_usage(account.name, cluster.name) -
+                     (get_cluster_limit(account.name, cluster.name)
+                      - active_sus - closing_allocations_query.aggregate(Sum("awarded"))))
+
+    # Close out any expired allocations
+    close_expired_allocations(closing_allocations_query.all(), current_usage)
+
+    # Gather the updated historical usage from expired allocations (including any newly expired allocations)
+    historical_usage = (Allocation.objects.filter(request__group=account,
+                                                  cluster=cluster,
+                                                  request__approved=True,
+                                                  request__expire__lte=date.today())
```

Comment:
Is there validation in the model that start dates need to be **before** end dates?

Reply:
No, but it would be good to add. I suggest leaving it for future work and raising an issue.